File: reference_impl/src/main/java/at/bumzack/reference/impl/ArticleService.java
package at.bumzack.reference.impl;

import at.bumzack.reference.impl.dto.Article;
import at.bumzack.reference.impl.dto.ArticleModel;
import at.bumzack.reference.impl.dto.ImportResult;
import at.bumzack.reference.impl.repository.ArticleRepository;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Component;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.math.BigDecimal;
import java.time.Instant;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.Objects;
import java.util.TimeZone;

import static java.util.Objects.nonNull;


@Component
public class ArticleService {
    private static final int LEN_CODE = 20;
    private static final int LEN_TITLE = 100;
    private static final int LEN_DESC = 1700;

    private static final int LEN_ATTRIBUTES = 200;

    private static final int LEN_CATEGORIES = 200;

    private static final int LEN_POS = 30;
    private static final int LEN_PRICE = 20;
    private static final int LEN_START = 25;
    // private static final int LEN_END = 25;


    public static final String PROPERTY_CODE = "code";
    private static final Logger LOG = LogManager.getLogger(ArticleService.class);
    private final ArticleRepository articleRepository;


    @Value("${sourcefilesFolder}")
    private String sourceFilesFolder;

    public ArticleService(final ArticleRepository articleRepository) {
        this.articleRepository = articleRepository;
    }

    public List<Article> findPaginated(final int pageNumber, final int pageSize) {
        final var page = PageRequest.of(pageNumber, pageSize, Sort.by(Sort.Direction.ASC, PROPERTY_CODE));
        return convert(articleRepository.findAll(page).toList());
    }

    private List<Article> convert(final List<ArticleModel> source) {
        return source.stream()
                .map(this::toArticle)
                .toList();
    }

    public ImportResult importArticles() {
        final var folder = new File(sourceFilesFolder);

        return Arrays.stream(Objects.requireNonNull(folder.listFiles()))
                .filter(file -> file.getName().contains(".txt"))
                .sorted(Comparator.comparing(File::getName))
                .map(this::tryProcessFile)
                .reduce(new ImportResult(), ImportResult::sum, ImportResult::sum);
    }

    public ImportResult importArticles2() {
        final var folder = new File(sourceFilesFolder);

        return Arrays.stream(Objects.requireNonNull(folder.listFiles()))
                .filter(file -> file.getName().contains(".txt"))
                .sorted(Comparator.comparing(File::getName))
                .toList()
                .parallelStream()
                .map(this::tryProcessFile)
                .reduce(new ImportResult(), ImportResult::sum, ImportResult::sum);
    }

    private ImportResult tryProcessFile(final File f) {
        try {
            final var res = processFile(f);
            LOG.info("filename {},  linesProcessed  {},   dbRowsWritten  {} ", f.getName(), res.getLinesProcessed(), res.getDbRowsWritten());
            return res;
        } catch (final IOException e) {
            LOG.error("error processing file ", e);
        }
        return null;
    }

    private ImportResult processFile(final File f) throws IOException {
        final var reader = new BufferedReader(new FileReader(f));
        long linesProcessed = 0;
        long dbRowsWritten = 0;

        String line = reader.readLine();
        linesProcessed++;

        final var article_grouped_by_code_and_pos = new ArrayList<ArticleModel>();
        final var articles_ready_to_write_to_db = new ArrayList<ArticleModel>();

        if (nonNull(line)) {
            var article = line2article(line);
            while (line != null) {
                LOG.info("line {},    article    code {}, pos {}, price  {}", linesProcessed,article.getCode(), article.getPos(), article.getPrice());
                final var firstInGroup = article;

                // read until code && pos change
                while (nonNull(line) && article.getCode().equals(firstInGroup.getCode()) && article.getPos().equals(firstInGroup.getPos())) {
                    article_grouped_by_code_and_pos.add(article);
                    line = reader.readLine();
                    linesProcessed++;
                    if (nonNull(line)) {
                        article = line2article(line);
                        LOG.info("line {},    article    code {}, pos {}, price  {}", linesProcessed,article.getCode(), article.getPos(), article.getPrice());
                    }
                }

                LOG.info("article finished grouping in line {},      code {}, pos {}, cnt_articles  {}",linesProcessed,firstInGroup.getCode(), firstInGroup.getPos(), article_grouped_by_code_and_pos.size());

                // find cheapest article
                final var cheapestArticle = article_grouped_by_code_and_pos.stream()
                        .sorted(Comparator.comparing(ArticleModel::getPrice))
                        .limit(1)
                        .toList();
                articles_ready_to_write_to_db.add(cheapestArticle.getFirst());

                // reset array and add the first article that wa
                article_grouped_by_code_and_pos.clear();
                if (line != null) {
                    article_grouped_by_code_and_pos.add(article);
                }

                if (articles_ready_to_write_to_db.size() > 0) {
                    // articleRepository.saveAll(articles);
                    dbRowsWritten += articles_ready_to_write_to_db.size();
                    articles_ready_to_write_to_db.forEach(a -> {
                        LOG.info("article ready to write to DB   code {}, pos {}", a.getCode(), a.getPos());
                    });
                    articles_ready_to_write_to_db.clear();
                }

                line = reader.readLine();
            }
        }

        final var importResult = new ImportResult();
        importResult.setDbRowsWritten(dbRowsWritten);
        importResult.setLinesProcessed(linesProcessed);
        return importResult;
    }


    private ArticleModel line2article(final String line) {
        final var article = new ArticleModel();
        int beginDesc = LEN_CODE + LEN_TITLE;
        int beginAttr = LEN_CODE + LEN_TITLE + LEN_DESC;
        int beginCat = LEN_CODE + LEN_TITLE + LEN_DESC + LEN_ATTRIBUTES;
        int beginPos = LEN_CODE + LEN_TITLE + LEN_DESC + LEN_ATTRIBUTES + LEN_CATEGORIES;
        int beginPrice = LEN_CODE + LEN_TITLE + LEN_DESC + LEN_ATTRIBUTES + LEN_CATEGORIES + LEN_POS;
        int beginStartDate = LEN_CODE + LEN_TITLE + LEN_DESC + LEN_ATTRIBUTES + LEN_CATEGORIES + LEN_POS + LEN_PRICE;
        int beginEndDate = LEN_CODE + LEN_TITLE + LEN_DESC + LEN_ATTRIBUTES + LEN_CATEGORIES + LEN_POS + LEN_PRICE + LEN_START;

        article.setCode(trimLeadingZeroes(line.substring(0, LEN_CODE)));
        article.setTitle(line.substring(LEN_CODE, beginDesc).trim());
        article.setDescription(line.substring(beginDesc, beginAttr).trim());
        article.setAttributes(line.substring(beginAttr, beginCat).trim());
        article.setCategories(line.substring(beginCat, beginPos).trim());
        article.setPos(trimLeadingZeroes(line.substring(beginPos, beginPrice).trim()));
        article.setPrice(BigDecimal.valueOf(Double.parseDouble(line.substring(beginPrice, beginStartDate))));
        article.setStartDate(LocalDateTime.from(LocalDateTime.ofInstant(Instant.ofEpochMilli(Long.parseLong(line.substring(beginStartDate, beginEndDate))),
                TimeZone.getDefault().toZoneId())));
        article.setEndDate(LocalDateTime.from(LocalDateTime.ofInstant(Instant.ofEpochMilli(Long.parseLong(line.substring(beginEndDate))),
                TimeZone.getDefault().toZoneId())));
        return article;
    }

    private String trimLeadingZeroes(final String s) {
        int i = 0;
        while (s.charAt(i) == '0') {
            i++;
        }
        return s.substring(i);
    }

    private Article toArticle(final ArticleModel article) {
        final var target = new Article();
        target.setAttributes(article.getAttributes());
        target.setCategories(article.getCategories());
        target.setCode(article.getCode());
        target.setDescription(article.getDescription());
        target.setId(article.getId());
        target.setPos(article.getPos());
        target.setTitle(article.getTitle());
        target.setStartDate(article.getStartDate().toString());
        target.setEndDate(article.getEndDate().toString());

        return target;
    }
}

File: vi-okr/gsc_graalvm_kotlin/src/main/kotlin/at/bumzack/refimpl/ArticleService.kt
package at.bumzack.refimpl

import at.bumzack.refimpl.dto.Article
import at.bumzack.refimpl.dto.ImportResult
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import java.io.BufferedReader
import java.io.File
import java.io.FileReader
import java.io.IOException
import java.math.BigDecimal
import java.time.Instant
import java.time.LocalDateTime
import java.util.*


private val LOG = LoggerFactory.getLogger(ArticleService::class.java)

private const val LEN_CODE = 20
private const val LEN_TITLE = 100
private const val LEN_DESC = 1700
private const val LEN_ATTRIBUTES = 200
private const val LEN_CATEGORIES = 200
private const val LEN_POS = 30
private const val LEN_PRICE = 20
private const val LEN_START = 25

@Service
class ArticleService {
    val LEN_CODE: Int = 20
    val LEN_TITLE: Int = 100
    val LEN_DESC: Int = 1700
    val LEN_ATTRIBUTES: Int = 200
    val LEN_CATEGORIES: Int = 200
    val LEN_POS: Int = 30
    val LEN_PRICE: Int = 20
    val LEN_START: Int = 25


    // private static final int LEN_END = 25;
    @Value("\${sourcefilesFolder}")
    private val sourceFilesFolder: String? = null


    fun importArticles(returnItems: Boolean): ImportResult {
        LOG.info("sourceFilesFolder {}", sourceFilesFolder)
        val folder = File(sourceFilesFolder!!)

        val res = Arrays.stream(Objects.requireNonNull(folder.listFiles()))
            .filter { file -> file.name.contains(".txt") }
            .sorted(Comparator.comparing { obj -> obj.name })
            .map { f -> tryProcessFile(f, returnItems) }
            .toList()

        var linesProcessed: Long = 0
        var dbRowsWritten: Long = 0

        val articles = mutableListOf<Article>()
        res.forEach { r ->
            linesProcessed += r?.linesProcessed ?: 0
            dbRowsWritten += r?.dbRowsWritten ?: 0
            articles.addAll(r?.articles ?: emptyList())
        }

        return ImportResult(
            linesProcessed = linesProcessed,
            dbRowsWritten = dbRowsWritten,
            articles = articles,
        )
    }

    fun importArticlesParallel(returnItems: Boolean): ImportResult {
        LOG.info("sourceFilesFolder {}", sourceFilesFolder)
        val folder = File(sourceFilesFolder!!)

        val res = Arrays.stream(Objects.requireNonNull(folder.listFiles()))
            .filter { file: File -> file.name.contains(".txt") }
            .sorted(Comparator.comparing { obj: File -> obj.name })
            .toList()
            .parallelStream()
            .map { f -> tryProcessFile(f, returnItems) }
            .toList()

        var linesProcessed: Long = 0
        var dbRowsWritten: Long = 0

        val articles = mutableListOf<Article>()
        res.forEach { r ->
            linesProcessed += r?.linesProcessed ?: 0
            dbRowsWritten += r?.dbRowsWritten ?: 0
            articles.addAll(r?.articles ?: emptyList())
        }

        return ImportResult(
            linesProcessed = linesProcessed,
            dbRowsWritten = dbRowsWritten,
            articles = articles,
        )
    }


    private fun tryProcessFile(f: File, returnItems: Boolean): ImportResult? {
        try {
            if (returnItems) {
                val res = processFile(f)
                LOG.info("filename {},  linesProcessed  {},   dbRowsWritten  {} ", f.name, res.linesProcessed, res.dbRowsWritten)
                return res
            } else {
                val res = processFileNoReturnItems(f)
                LOG.info("filename {},  linesProcessed  {},   dbRowsWritten  {} ", f.name, res.linesProcessed, res.dbRowsWritten)
                return res
            }
        } catch (e: IOException) {
            LOG.error("error processing file ", e)
        }
        return null
    }

    @Throws(IOException::class)
    private fun processFile(f: File): ImportResult {
        val reader = BufferedReader(FileReader(f))
        var linesProcessed: Long = 0
        var dbRowsWritten: Long = 0

        var line = reader.readLine()
        linesProcessed++

        val article_grouped_by_code_and_pos = ArrayList<Article>()
        val articles_ready_to_write_to_db = ArrayList<Article>()

        if (Objects.nonNull(line)) {
            var article: Article
            var prevArticle: Article? = null
            while (true) {
                article = line2article(line)

                if (Objects.isNull(prevArticle)) {
                    // new grouping start - because first article ever
                    article_grouped_by_code_and_pos.add(article)
                } else {
                    // is article part of current group?
                    if (article.code == prevArticle?.code && article.pos == prevArticle.pos) {
                        article_grouped_by_code_and_pos.add(article)
                    } else {
                        // article is not part of current group -> find cheapest
                        val cheapestArticle = article_grouped_by_code_and_pos.stream()
                            .sorted(Comparator.comparing(Article::price))
                            .limit(1)
                            .toList()
                        // comment for "returnItems == false"
                        articles_ready_to_write_to_db.add(cheapestArticle.first())

                        dbRowsWritten++

                        // clear group and add article
                        article_grouped_by_code_and_pos.clear()
                        article_grouped_by_code_and_pos.add(article)
                    }
                }

                line = reader.readLine()
                if (Objects.isNull(line)) {
                    break
                }
                linesProcessed++
                prevArticle = article
            }

            // write last article in file
            val cheapestArticle = article_grouped_by_code_and_pos.stream()
                .sorted(Comparator.comparing(Article::price))
                .limit(1)
                .toList()

            // comment for "returnItems == false"
            articles_ready_to_write_to_db.add(cheapestArticle.first())

            dbRowsWritten++
        }

        val importResult = ImportResult()
        importResult.dbRowsWritten = dbRowsWritten
        importResult.linesProcessed = linesProcessed
        importResult.articles = articles_ready_to_write_to_db

        return importResult
    }

    @Throws(IOException::class)
    private fun processFileNoReturnItems(f: File): ImportResult {
        val reader = BufferedReader(FileReader(f))
        var linesProcessed: Long = 0
        var dbRowsWritten: Long = 0

        var line = reader.readLine()
        linesProcessed++

        val article_grouped_by_code_and_pos = ArrayList<Article>()
        val articles_ready_to_write_to_db = ArrayList<Article>()

        if (Objects.nonNull(line)) {
            var article: Article
            var prevArticle: Article? = null
            while (true) {
                article = line2article(line)

                if (Objects.isNull(prevArticle)) {
                    // new grouping start - because first article ever
                    article_grouped_by_code_and_pos.add(article)
                } else {
                    // is article part of current group?
                    if (article.code.equals(prevArticle?.code) && article.pos.equals(prevArticle?.pos)) {
                        article_grouped_by_code_and_pos.add(article)
                    } else {
                        // article is not part of current group -> find cheapest
                        val cheapestArticle = article_grouped_by_code_and_pos.stream()
                            .sorted(Comparator.comparing(Article::price))
                            .limit(1)
                            .toList()

                        // comment for "returnItems == false"
                        // articles_ready_to_write_to_db.add(cheapestArticle.getFirst());
                        dbRowsWritten++

                        // clear group and add article
                        article_grouped_by_code_and_pos.clear()
                        article_grouped_by_code_and_pos.add(article)
                    }
                }

                line = reader.readLine()
                if (Objects.isNull(line)) {
                    break
                }
                linesProcessed++
                prevArticle = article
            }

            // write first article in list to response
            val cheapestArticle = article_grouped_by_code_and_pos.stream()
                .sorted(Comparator.comparing(Article::price))
                .limit(1)
                .toList()
            // comment for "returnItems == false"
//            articles_ready_to_write_to_db.add(cheapestArticle.getFirst());
            dbRowsWritten++
        }

        val importResult = ImportResult()
        importResult.dbRowsWritten = dbRowsWritten
        importResult.linesProcessed = linesProcessed
        importResult.articles = articles_ready_to_write_to_db

        return importResult
    }

    private fun line2article(line: String): Article {

        val beginDesc = LEN_CODE + LEN_TITLE
        val beginAttr = LEN_CODE + LEN_TITLE + LEN_DESC
        val beginCat = LEN_CODE + LEN_TITLE + LEN_DESC + LEN_ATTRIBUTES
        val beginPos = LEN_CODE + LEN_TITLE + LEN_DESC + LEN_ATTRIBUTES + LEN_CATEGORIES
        val beginPrice = LEN_CODE + LEN_TITLE + LEN_DESC + LEN_ATTRIBUTES + LEN_CATEGORIES + LEN_POS
        val beginStartDate = LEN_CODE + LEN_TITLE + LEN_DESC + LEN_ATTRIBUTES + LEN_CATEGORIES + LEN_POS + LEN_PRICE
        val beginEndDate = LEN_CODE + LEN_TITLE + LEN_DESC + LEN_ATTRIBUTES + LEN_CATEGORIES + LEN_POS + LEN_PRICE + LEN_START


        val startDateStr = line.substring(beginStartDate, beginEndDate)
        val endDateStr = line.substring(beginEndDate)
        val start = LocalDateTime.ofInstant(
            Instant.ofEpochSecond(startDateStr.toLong()),
            TimeZone.getDefault().toZoneId()
        )
        val end = LocalDateTime.ofInstant(
            Instant.ofEpochSecond(endDateStr.toLong()),
            TimeZone.getDefault().toZoneId()
        )

        val article = Article(
            code = trimLeadingZeroes(line.substring(0, LEN_CODE)),
            title = line.substring(LEN_CODE, beginDesc).trim { it <= ' ' },
            description = line.substring(beginDesc, beginAttr).trim { it <= ' ' },
            attributes = line.substring(beginAttr, beginCat).trim { it <= ' ' },
            categories = line.substring(beginCat, beginPos).trim { it <= ' ' },
            pos = trimLeadingZeroes(line.substring(beginPos, beginPrice).trim { it <= ' ' }),
            price = BigDecimal.valueOf(line.substring(beginPrice, beginStartDate).toDouble()),
            startDate = LocalDateTime.from(start).toString(), endDate = LocalDateTime.from(end).toString()
        )

        return article
    }

    private fun trimLeadingZeroes(s: String): String {
        var i = 0
        while (s[i] == '0') {
            i++
        }
        return s.substring(i)
    }
}
